ee/tabby-webserver/src/schema/email.rs

```diff
@@ -1,11 +1,120 @@
 use anyhow::Result;
 use async_trait::async_trait;
-use juniper::GraphQLObject;
+use juniper::{GraphQLEnum, GraphQLObject};
+use serde::Serialize;
+use validator::{validate_url, ValidationError};
+
+#[derive(GraphQLEnum, Clone, Debug)]
+pub enum Encryption {
+    StartTls,
+    SslTls,
+    None,
+}
+
+impl ToString for Encryption {
+    fn to_string(&self) -> String {
+        match self {
+            Encryption::StartTls => "STARTTLS",
+            Encryption::SslTls => "SSLTLS",
+            Encryption::None => "NONE",
+        }
+        .into()
+    }
+}
+
+#[derive(GraphQLEnum, Clone, Debug)]
+pub enum AuthMethod {
+    Plain,
+    Login,
+    XOAuth2,
+}
+
+impl ToString for AuthMethod {
+    fn to_string(&self) -> String {
+        match self {
+            AuthMethod::Plain => "PLAIN",
+            AuthMethod::Login => "LOGIN",
+            AuthMethod::XOAuth2 => "XOAUTH2",
+        }
+        .into()
+    }
+}
 
 #[derive(GraphQLObject)]
 pub struct EmailSetting {
     pub smtp_username: String,
     pub smtp_server: String,
+    pub from_address: String,
+    pub encryption: Encryption,
+    pub auth_method: AuthMethod,
+}
+
+fn validation_error<T: Serialize>(
+    msg: &'static str,
+    field: &'static str,
+    value: &T,
+) -> ValidationError {
+    let mut err = ValidationError::new(msg);
+    err.add_param(field.into(), value);
+    err
+}
+
+impl EmailSetting {
+    pub fn new_validate(
+        smtp_username: String,
+        smtp_server: String,
+        from_address: Option<String>,
+        encryption: String,
+        auth_method: String,
+    ) -> Result<Self, ValidationError> {
+        if !validate_url(&smtp_server) {
+            return Err(validation_error(
+                "Invalid server URL",
+                "smtp_server",
+                &smtp_server,
+            ));
+        }
+
+        let encryption = Self::convert_encryption(&encryption)?;
+        let auth_method = Self::convert_auth_method(&auth_method)?;
+
+        let from_address = from_address.unwrap_or_else(|| smtp_username.clone());
+        Ok(EmailSetting {
+            smtp_username,
+            smtp_server,
+            from_address,
+            encryption,
+            auth_method,
+        })
+    }
+
+    pub fn convert_encryption(encryption: &str) -> Result<Encryption, ValidationError> {
```

Comment:
I suggest just use serde serialize / deserialize to parse / serialize the enums. (though still return an anyhow error if deserialize failed)

Reply:
I don't think using serde here really makes sense. We would need to implement a custom Serializer/Deserializer. I considered using strum as well, but pulling that in just for this also seems overkill. I think it's fine to have a few string-based match statements for now until we have enough of them that it warrants conversion to strum, which I think we should prefer over serde for direct string <-> enum mapping.